google/cloud/pubsublite/internal/alarm_registry_impl_test.cc

```diff
@@ -0,0 +1,177 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/pubsublite/internal/alarm_registry_impl.h"
+#include "google/cloud/future.h"
+#include "google/cloud/status_or.h"
+#include "google/cloud/testing_util/mock_completion_queue_impl.h"
+#include <gmock/gmock.h>
+#include <chrono>
+#include <deque>
+#include <memory>
+#include <thread>
+
+namespace google {
+namespace cloud {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace pubsublite_internal {
+namespace {
+
+using google::cloud::CompletionQueue;
+using google::cloud::testing_util::MockCompletionQueueImpl;
+using ::testing::ByMove;
+using ::testing::InSequence;
+;
+using ::testing::MockFunction;
+using ::testing::Return;
+using ::testing::StrictMock;
+
+// 10,000 seconds to ensure this test has no time dependence
+std::chrono::milliseconds constexpr kAlarmPeriod{10000 * 1000};
+
+class AlarmRegistryImplTest : public ::testing::Test {
+ protected:
+  AlarmRegistryImplTest()
+      : cq_{std::make_shared<StrictMock<MockCompletionQueueImpl>>()},
+        alarm_{CompletionQueue{cq_}} {}
+  std::shared_ptr<StrictMock<MockCompletionQueueImpl>> cq_;
+  AlarmRegistryImpl alarm_;
+  StrictMock<MockFunction<void()>> fun_;
+};
+
+TEST_F(AlarmRegistryImplTest, TokenDestroyedBeforeRun) {
+  promise<StatusOr<std::chrono::system_clock::time_point>> timer;
+  EXPECT_CALL(*cq_, MakeRelativeTimer(std::chrono::nanoseconds(kAlarmPeriod)))
+      .WillOnce(Return(ByMove(timer.get_future())));
+  auto token = alarm_.RegisterAlarm(kAlarmPeriod, fun_.AsStdFunction());
+  token = nullptr;
+  timer.set_value(std::chrono::system_clock::time_point{});
+}
+
+TEST_F(AlarmRegistryImplTest, TimerErrorBeforeRun) {
+  promise<StatusOr<std::chrono::system_clock::time_point>> timer;
+  EXPECT_CALL(*cq_, MakeRelativeTimer(std::chrono::nanoseconds(kAlarmPeriod)))
+      .WillOnce(Return(ByMove(timer.get_future())));
+  auto token = alarm_.RegisterAlarm(kAlarmPeriod, fun_.AsStdFunction());
+  timer.set_value(Status{StatusCode::kCancelled, "cancelled"});
+}
+
+TEST_F(AlarmRegistryImplTest, TokenDestroyedAfterSingleRun) {
+  InSequence seq;
+
+  promise<StatusOr<std::chrono::system_clock::time_point>> timer;
+  EXPECT_CALL(*cq_, MakeRelativeTimer(std::chrono::nanoseconds(kAlarmPeriod)))
+      .WillOnce(Return(ByMove(timer.get_future())));
+  auto token = alarm_.RegisterAlarm(kAlarmPeriod, fun_.AsStdFunction());
+
+  EXPECT_CALL(fun_, Call);
+
+  auto temp_promise = std::move(timer);
+  timer = promise<StatusOr<std::chrono::system_clock::time_point>>{};
```

Comment:
just do `promise<StatusOr<std::chrono::system_clock::time_point>> timer2;` instead of moving out of timer